pkg/login/social/google_oauth.go

```diff
@@ -11,35 +11,103 @@ import (
 	"github.com/grafana/grafana/pkg/services/featuremgmt"
 )
 
+const legacyAPIURL = "https://www.googleapis.com/oauth2/v1/userinfo"
+const googleIAMGroupsEndpoint = "https://content-cloudidentity.googleapis.com/v1/groups/-/memberships:searchDirectGroups"
+
 type SocialGoogle struct {
 	*SocialBase
 	hostedDomain string
 	apiUrl       string
 }
 
+type googleUserData struct {
+	ID            string `json:"sub"`
+	Email         string `json:"email"`
+	Name          string `json:"name"`
+	EmailVerified bool   `json:"email_verified"`
+}
+
 func (s *SocialGoogle) UserInfo(ctx context.Context, client *http.Client, token *oauth2.Token) (*BasicUserInfo, error) {
-	var data struct {
-		Id    string `json:"id"`
-		Name  string `json:"name"`
-		Email string `json:"email"`
+	data, errToken := s.extractFromToken(ctx, client, token)
+	if errToken != nil {
+		return nil, errToken
+	}
+
+	if data == nil {
+		var errAPI error
+		data, errAPI = s.extractFromAPI(ctx, client)
+		if errAPI != nil {
+			return nil, errAPI
+		}
+	}
+
+	if data.ID == "" {
+		return nil, fmt.Errorf("error getting user info: id is empty")
+	}
+
+	if !data.EmailVerified {
+		return nil, fmt.Errorf("user email is not verified")
+	}
+
+	groups, errPage := s.retrieveGroups(ctx, client, data)
+	if errPage != nil {
+		s.log.Warn("Error retrieving groups", "error", errPage)
+	}
+
+	userInfo := &BasicUserInfo{
+		Id:             data.ID,
+		Name:           data.Name,
+		Email:          data.Email,
+		Login:          data.Email,
+		Role:           "",
+		IsGrafanaAdmin: nil,
+		Groups:         groups,
+	}
+
+	s.log.Debug("Resolved user info", "data", fmt.Sprintf("%+v", userInfo))
+
+	return userInfo, nil
+}
+
+type googleAPIData struct {
+	ID            string `json:"id"`
+	Name          string `json:"name"`
+	Email         string `json:"email"`
+	EmailVerified bool   `json:"verified_email"`
+}
+
+func (s *SocialGoogle) extractFromAPI(ctx context.Context, client *http.Client) (*googleUserData, error) {
+	if legacyAPIURL == s.apiUrl {
```

Comment:
thought: should we check that legacyAPIURL is the prefix of `s.apiUrl`? In case the user has specified `https://www.googleapis.com/oauth2/v1/userinfo/` as the URL.